pysensors/basis/_custom.py

```diff
@@ -0,0 +1,82 @@
+"""
+custom mode basis class.
+"""
+from ._base import InvertibleBasis
+from ._base import MatrixMixin
+
+class Custom(InvertibleBasis, MatrixMixin):
+    """
+    Generate a custom transformation which maps input features to
+    custom modes.
+
+    Assumes the data has already been centered (to have mean 0).
+
+    Parameters
+    ----------
+    n_basis_modes : int, optional (default 10)
+        Number of basis modes to retain. Cannot be larger than
+        the number of features ``n_features``, or the number of examples
+        ``n_examples``.
+
+    Attributes
+    ----------
+    basis_matrix_ : numpy ndarray, shape (n_features, n_basis_modes)
+        The top n_basis_modes left singular vectors of the training data.
+
+    """
+
+    def __init__(self, n_basis_modes=10,U = None, **kwargs):
+        if isinstance(n_basis_modes, int) and n_basis_modes > 0:
+            super(Custom, self).__init__()#n_components=n_basis_modes, **kwargs
+            self._n_basis_modes = n_basis_modes
+            self.custom_basis_ = U
+        else:
+            raise ValueError("n_basis_modes must be a positive integer.")
+
+    def fit(self, X):
+        """
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, n_features)
+            The training data.
+
+        Returns
+        -------
+        self : instance
+        """
+        # self.basis_matrix_ = super(Custom, self).fit(X).components_.T
+        self.basis_matrix_ = self.custom_basis_
+        return self
+
+    def matrix_inverse(self, n_basis_modes=None):
+        """
+        Get the inverse matrix mapping from measurement space to
+        coordinates with respect to the basis.
+
+        Note that this is not the inverse of the matrix returned by
+        ``self.matrix_representation``. It is the (psuedo) inverse of
+        the matrix whose columns are the basis modes.
+
+        Parameters
+        ----------
+        n_basis_modes : positive int, optional (default None)
+            Number of basis modes to be used to compute inverse.
+
+        Returns
+        -------
+        B : numpy ndarray, shape (n_basis_modes, n_features)
+            The inverse matrix.
+        """
+        n_basis_modes = self._validate_input(n_basis_modes)
+
+        return self.basis_matrix_[:, :n_basis_modes].T
+
+    @property
+    def n_basis_modes(self):
+        """Number of basis modes."""
+        return self._n_basis_modes
+
+    @n_basis_modes.setter
+    def n_basis_modes(self, n_basis_modes):
+        self._n_basis_modes = n_basis_modes
```

Comment:
Unless  ```@n_basis_modes.setter``` is required, I would recommend not including it and the rest of this function.  It can let people make the custom basis inconsistent with the _n_basis_modes.

Reply:
I think that's the whole idea, they want to loop over n_basis modes and see which is the best value. but we should have an error if the n_basis is > U.shape[1]. I think this already exists in the parent class but I need to double-check.

Reply:
Okay.